text/2020-04-27-encryption-at-rest.md

```diff
@@ -0,0 +1,195 @@
+
+The actual encryption happens both in storage engine (i.e. RocksDB) and out side of storage engine (for temporary files used by TiKV, like snapshot files and imported backup files). In either of the cases, OpenSSL is used to encrypt the files and the encryption format are compatible with each other. This means, for example, SST files generated and encrypted outside of RocksDB by TiKV are able to ingest into RocksDB as is, as long as the key manager module is able to pass encryption info along correctly.
+
+### Providing and Rotating Master Key
+
+Management of master key is external to TiKV. Master key can be provided to TiKV in one of the following ways. We need to update config format accordingly.
+
+* Managed in AWS Key Management Service (KMS, https://aws.amazon.com/kms/). Currently we only support AWS KMS.
+* Pass through a file, whose path is specified in TiKV config.
+
+KMS is the preferred way for master key management in production. To use it, user would need to specify the KMS CMK to use in TiKV config. There are two APIs that TiKV uses from KMS service: `GenerateDataKey` and `Decrypt`. The `GenerateDataKey` is capable of generating a new encryption key and return both of its plaintext and ciphertext form. When encrypted TiKV data keys, TiKV will call `GenerateDataKey` to generate a KMS data key (not to confuse with TiKV data key) and use it to encrypt TiKV data keys, and store the ciphertext key along with the TiKV data keys. When TiKV need to load TiKV data keys from disk, TiKV will call the `Decrypt` API the decrypt the ciphertext key, and use the result as the key to decrypt the TiKV data keys.
+
+Note the extra layer of envolope encryption used here (KMS CMK -> KMS data key -> TiKV data key). We don't use KMS data key directly as TiKV data key is because AWS KMS currently doesn't provide a way for bulk decrypt. If we accumulate a number of data keys in the system (think of by default we rotate data key every week, after a year there will be 50+ data keys co-existing), on TiKV restart we don't want to send multiple requests to AWS to decrypt them one by one.
+
+Alternatively passing master key stored in a file is also supported, but it is not recommended to use in production. In that case the plaintext key is used directly to encrypt and decrypt (TiKV) data keys.
+
+In both of the cases, AES256-GCM is used to encrypt the (TiKV) data keys. The main reason to use GCM mode instead of CTR mode is not to verify authenticity of the persisted data keys, but rather provide a way to verify if the right master key is provided to decrypt the key. Without using an authentication encryption method like GCM, we don't have reliable way to tell if user provide a different key than the one use previously to encrypt the data keys. The use of AES256 here is non-configurable, in order to reduce configurable knobs that could confuse user.
```

Comment:
```suggestion
In both of the cases, AES256-GCM is used to encrypt the (TiKV) data keys. The main reason to use GCM mode instead of CTR mode is not to verify authenticity of the persisted data keys, but rather provide a way to verify if the right master key is provided to decrypt the key. Without using an authentication encryption method like GCM, we don't have a reliable way to tell if user provides a different key than the one used previously to encrypt the data keys. The use of AES256 here is non-configurable, in order to reduce configurable knobs that could confuse user.
```